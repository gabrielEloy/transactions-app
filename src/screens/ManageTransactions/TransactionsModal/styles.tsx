import {StyleSheet} from 'react-native';
import {COLORS} from 'src/themes/main';

const styles = StyleSheet.create({
  modalStyle: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: 'rgba(0,0,0,.3)',
    borderRadius: 10,
  },
  modalAction: {
    width: '80%',
    height: 450,
    backgroundColor: COLORS.background,
    borderRadius: 20,
    overflow: 'hidden',
  },
  modalHeader: {
    position: 'relative',
    justifyContent: 'center',
    alignItems: 'center',
    paddingVertical: 20,
  },
  headerText: {
    fontSize: 18,
    color: COLORS.light,
  },
  footer: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'flex-end',
    alignItems: 'flex-end',
    marginRight: 25,
    marginBottom: 20,
  },
  text: {
    fontFamily: 'Exo-Medium',
    color: COLORS.light,
  },
  form: {
    flex: 2,
    padding: 20,
  },
  input: {
    borderBottomWidth: 1,
    borderBottomColor: COLORS.gray,
    color: COLORS.light,
    fontFamily: 'Exo-Medium',
  },
  inputContainer: {
    marginTop: 20,
  },
  switchStyle: {
    borderRadius: 0,
    marginTop: 20,
  },
  blackText: {color: 'black'},
});

export default styles;
