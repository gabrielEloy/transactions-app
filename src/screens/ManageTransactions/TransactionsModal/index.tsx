import React, {useState, useContext} from 'react';
import {View, Text, Modal, TextInput, Alert} from 'react-native';
import SwitchSelector from 'react-native-switch-selector';

import {COLORS} from 'src/themes/main';
import Button from 'src/components/Button';
import Divider from 'src/components/Divider';
import TransactionsContext from 'src/contexts/TransactionsContext';
import CurrencyInput from 'src/components/CurrencyInput';
import {getValueAndCurrency} from 'src/helpers/currency';
import styles from './styles';

interface Props {
  isModalVisible: boolean;
}

const TransactionsModal = (props: Props) => {
  const [description, setDescription] = useState('');
  const [value, setValue] = useState('');
  const [debit, setIsDebit] = useState(false);

  const context = useContext(TransactionsContext);

  const resetState = () => {
    setValue('');
    setDescription('');
    setIsDebit(false);
  };

  const handleSwitchSelect = (switchValue: string) => {
    setIsDebit(switchValue === 'Debit');
  };
  const handleValue = (val: string) => {
    setValue(val);
  };
  const handleDescription = (descValue: string) => setDescription(descValue);

  const checkFilledFields = () => {
    let errors = '';

    if (!description) {
      errors += '- missing description \n';
    }
    if (!value) {
      errors += '- missing value \n';
    }

    return errors;
  };

  const handleAddTransaction = () => {
    const errors = checkFilledFields();

    if (errors) {
      Alert.alert('Missing fields', errors);
      return;
    }

    context.addTransaction({
      description,
      value: getValueAndCurrency(value),
      debit,
    });
    context.setModalVisible(false);
    resetState();
  };

  const cancelAddTransaction = () => {
    resetState();
    context.setModalVisible(false);
  };

  const options = [
    {label: 'Credit', value: 'Credit', activeColor: COLORS.primary},
    {label: 'Debit', value: 'Debit', activeColor: COLORS.primary},
  ];

  return (
    <Modal transparent animationType="fade" visible={props.isModalVisible}>
      <View style={styles.modalStyle}>
        <View style={styles.modalAction}>
          <View style={styles.modalHeader}>
            <Text style={styles.headerText}>New Transaction</Text>
            <Divider />
          </View>
          <View style={styles.form}>
            <View>
              <Text style={styles.text}>Type</Text>
              <SwitchSelector
                initial={0}
                borderRadius={5}
                animationDuration={200}
                style={styles.switchStyle}
                options={options}
                onPress={handleSwitchSelect}
                textStyle={{...styles.text, ...styles.blackText}}
              />
            </View>
            <View style={styles.inputContainer}>
              <Text style={styles.text}>Description</Text>
              <TextInput
                testID="descriptionInput"
                value={description}
                onChangeText={handleDescription}
                style={styles.input}
              />
            </View>
            <View style={styles.inputContainer}>
              <Text style={styles.text}>Value</Text>
              <CurrencyInput
                testID="valueInput"
                value={value}
                onChangeText={handleValue}
                style={styles.input}
              />
            </View>
          </View>
          <View style={styles.footer}>
            <Button
              testID="cancelButton"
              onPress={cancelAddTransaction}
              removeBackground>
              <Text style={styles.text}>Cancel</Text>
            </Button>
            <Button
              testID="addTransactionButton"
              onPress={handleAddTransaction}>
              <Text style={styles.text}>Add</Text>
            </Button>
          </View>
        </View>
      </View>
    </Modal>
  );
};

export default TransactionsModal;
