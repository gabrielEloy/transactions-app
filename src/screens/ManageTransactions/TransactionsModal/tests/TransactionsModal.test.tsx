import React from 'react';
import {render, fireEvent} from 'react-native-testing-library';
import TransactionsModal from '../index';

describe('test <TransactionsModal />', () => {
  test('should successfully render component', () => {
    const tree = render(<TransactionsModal isModalVisible />);
    expect(tree).toMatchSnapshot();
  });

  test('Should correctly render the component inputs', async () => {
    const {getByTestId, findByText} = render(
      <TransactionsModal isModalVisible />,
    );

    const descriptionInput = getByTestId('descriptionInput');
    const valueInput = getByTestId('valueInput');

    fireEvent.changeText(descriptionInput, 'received money');
    fireEvent.changeText(valueInput, '22,50');

    expect(findByText('received money')).not.toBeNull();
    expect(findByText('R$ 22,50')).not.toBeNull();
  });

  test('Should correctly render the component inputs', async () => {
    const {getByTestId, findByText} = render(
      <TransactionsModal isModalVisible />,
    );

    const descriptionInput = getByTestId('descriptionInput');
    const valueInput = getByTestId('valueInput');

    fireEvent.changeText(descriptionInput, 'received money');
    fireEvent.changeText(valueInput, '22,50');

    expect(findByText('received money')).not.toBeNull();
    expect(findByText('R$ 22,50')).not.toBeNull();
  });

  test('Nothing should happen when ', () => {
    const {getByTestId} = render(<TransactionsModal isModalVisible />);

    const addTransactionButton = getByTestId('addTransactionButton');
    fireEvent.press(addTransactionButton);

    const descriptionInput = getByTestId('descriptionInput');
    const valueInput = getByTestId('valueInput');

    expect(descriptionInput.props.value).toBeFalsy();
    expect(valueInput.props.value).toBeFalsy();
  });
});
