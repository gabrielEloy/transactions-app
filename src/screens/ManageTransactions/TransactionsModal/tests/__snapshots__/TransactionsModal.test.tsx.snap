// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test <TransactionsModal /> should successfully render component 1`] = `
<Modal
  animationType="fade"
  hardwareAccelerated={false}
  transparent={true}
  visible={true}
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "rgba(0,0,0,.3)",
        "borderRadius": 10,
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#2D2D38",
          "borderRadius": 20,
          "height": 450,
          "overflow": "hidden",
          "width": "80%",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "justifyContent": "center",
            "paddingVertical": 20,
            "position": "relative",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#FBFBFB",
              "fontSize": 18,
            }
          }
        >
          New Transaction
        </Text>
        <View
          style={
            Array [
              Object {
                "bottom": 0,
                "left": 0,
                "position": "absolute",
                "right": 0,
              },
              Object {
                "backgroundColor": "#7A7A80",
                "height": 1,
              },
            ]
          }
        />
      </View>
      <View
        style={
          Object {
            "flex": 2,
            "padding": 20,
          }
        }
      >
        <View>
          <Text
            style={
              Object {
                "color": "#FBFBFB",
                "fontFamily": "Exo-Medium",
              }
            }
          >
            Type
          </Text>
          <View
            style={
              Array [
                Object {
                  "flexDirection": "row",
                },
                Object {
                  "borderRadius": 0,
                  "marginTop": 20,
                },
              ]
            }
          >
            <View
              onMoveShouldSetResponder={[Function]}
              onMoveShouldSetResponderCapture={[Function]}
              onResponderEnd={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderReject={[Function]}
              onResponderRelease={[Function]}
              onResponderStart={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              onStartShouldSetResponderCapture={[Function]}
              style={
                Object {
                  "flex": 1,
                }
              }
            >
              <View
                onLayout={[Function]}
                style={
                  Object {
                    "backgroundColor": "#FFFFFF",
                    "borderRadius": 5,
                    "height": 40,
                  }
                }
              >
                <View
                  style={
                    Object {
                      "alignItems": "center",
                      "borderColor": "#C9C9C9",
                      "borderRadius": 5,
                      "borderWidth": 0,
                      "flex": 1,
                      "flexDirection": "row",
                    }
                  }
                >
                  <View
                    accessible={true}
                    focusable={true}
                    onClick={[Function]}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      Object {
                        "alignItems": "center",
                        "flex": 1,
                        "flexDirection": "row",
                        "justifyContent": "center",
                        "opacity": 1,
                      }
                    }
                  >
                    <Text
                      style={
                        Array [
                          Object {
                            "backgroundColor": "transparent",
                            "color": "#FFFFFF",
                            "fontSize": 14,
                            "fontWeight": "normal",
                            "textAlign": "center",
                          },
                          Object {},
                        ]
                      }
                    >
                      Credit
                    </Text>
                  </View>
                  <View
                    accessible={true}
                    focusable={true}
                    onClick={[Function]}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      Object {
                        "alignItems": "center",
                        "flex": 1,
                        "flexDirection": "row",
                        "justifyContent": "center",
                        "opacity": 1,
                      }
                    }
                  >
                    <Text
                      style={
                        Array [
                          Object {
                            "backgroundColor": "transparent",
                            "color": "#000000",
                            "fontSize": 14,
                            "fontWeight": "normal",
                            "textAlign": "center",
                          },
                          Object {
                            "color": "black",
                            "fontFamily": "Exo-Medium",
                          },
                        ]
                      }
                    >
                      Debit
                    </Text>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
        <View
          style={
            Object {
              "marginTop": 20,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#FBFBFB",
                "fontFamily": "Exo-Medium",
              }
            }
          >
            Description
          </Text>
          <TextInput
            allowFontScaling={true}
            onChangeText={[Function]}
            rejectResponderTermination={true}
            style={
              Object {
                "borderBottomColor": "#7A7A80",
                "borderBottomWidth": 1,
                "color": "#FBFBFB",
                "fontFamily": "Exo-Medium",
              }
            }
            testID="descriptionInput"
            underlineColorAndroid="transparent"
            value=""
          />
        </View>
        <View
          style={
            Object {
              "marginTop": 20,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#FBFBFB",
                "fontFamily": "Exo-Medium",
              }
            }
          >
            Value
          </Text>
          <TextInput
            allowFontScaling={true}
            keyboardType="numeric"
            onChangeText={[Function]}
            options={
              Object {
                "delimiter": ".",
                "precision": 2,
                "separator": ",",
                "suffixUnit": "",
                "unit": "R$ ",
              }
            }
            rejectResponderTermination={true}
            style={
              Object {
                "borderBottomColor": "#7A7A80",
                "borderBottomWidth": 1,
                "color": "#FBFBFB",
                "fontFamily": "Exo-Medium",
              }
            }
            testID="valueInput"
            type="money"
            underlineColorAndroid="transparent"
            value=""
          />
        </View>
      </View>
      <View
        style={
          Object {
            "alignItems": "flex-end",
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "flex-end",
            "marginBottom": 20,
            "marginRight": 25,
          }
        }
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "backgroundColor": "rgba(0,0,0,0)",
              "borderRadius": 5,
              "opacity": 1,
              "paddingHorizontal": 20,
              "paddingVertical": 15,
            }
          }
          testID="cancelButton"
        >
          <Text
            style={
              Object {
                "color": "#FBFBFB",
                "fontFamily": "Exo-Medium",
              }
            }
          >
            Cancel
          </Text>
        </View>
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "backgroundColor": "#903EE3",
              "borderRadius": 5,
              "opacity": 1,
              "paddingHorizontal": 20,
              "paddingVertical": 15,
            }
          }
          testID="addTransactionButton"
        >
          <Text
            style={
              Object {
                "color": "#FBFBFB",
                "fontFamily": "Exo-Medium",
              }
            }
          >
            Add
          </Text>
        </View>
      </View>
    </View>
  </View>
</Modal>
`;
